[GSOC Taks 2] ceph-devstack Cli  logs feature : To Display Teuthology Logs

abhi4578 opened the pull request:
## [Adds implementation of logs and --logs-file option](https://github.com/zmc/ceph-devstack/commit/43bf143a3713b021e9ddb9fa08696d441a9b850d)
- ``ceph-devstack logs `` displays latest run
teuthology logs of given job
- ``ceph-devstack logs --log-file `` displays
file location of latest run teuthology log of
given job

## [Adds pytest unit tests to test 'logs' command](https://github.com/zmc/ceph-devstack/commit/5e200303fe97a95228d38bfcb1a4e199977b60ee)
- Pytest unit tests are added to test new logs command with different 
 test cases using parametrized tests and fixtures

abhi4578 committed:
Add logs and --log-file option in parse_args

abhi4578 committed:
Add implementation of logs and --logs-file option

- ``ceph-devstack logs `` displays latest run
teuthology logs of given job
- ``ceph-devstack logs --log-file `` displace
file location of latest run teuthology log of
given job

abhi4578 committed:
Add type hints to teuthology function and comments

abhi4578 committed:
Adding pytest unit tests to test 'logs' command

abhi4578 commented:
Hello @amathuria @VallariAg @zmc  here is my gsoc task 2  logs command implementation on ceph-devstack. Looking forward to your feedback.

abhi4578 committed:
Update README with 'logs' command

abhi4578 commented:
Disclosure: Github Copilot was used to add doc-strings (like args of function docs/comments) and few one liner python exceptions. Please let me know if this is acceptable or needs to be changed accordingly?

abhi4578 committed:
Update README.md

Signed-off-by: Abhi  <29863835+abhi4578@users.noreply.github.com>

kamoltat reviewed (commented):
Reviewing

kamoltat commented on ceph_devstack/resources/test/test_logs.py:
```diff
@@ -0,0 +1,101 @@
+    """
+    logger = logging.getLogger(__name__)
+
+    if flag_set=="true":
```
```suggestion
    if flag_set:
```

kamoltat commented on ceph_devstack/cli.py:
```diff
@@ -44,3 +49,51 @@ async def run():
+    if len(list_runs) == 0:
+        logger.error("No runs found!")
+        return 1  
+    list_runs.sort(key=lambda x: datetime.strptime(x.split('/')[-1].split('root-')[1].split('-teuthology')[0], '%Y-%m-%d_%H:%M:%S'))
```
pretty risky, consider using try and except for IndexError, ValueError.

abhi4578 replied:
Done.

kamoltat commented on ceph_devstack/cli.py:
```diff
@@ -44,3 +49,51 @@ async def run():
+        logger.error("No runs found!")
+        return 1  
+    list_runs.sort(key=lambda x: datetime.strptime(x.split('/')[-1].split('root-')[1].split('-teuthology')[0], '%Y-%m-%d_%H:%M:%S'))
+    latest_run = list_runs[-1]
```
maybe a comment that we can index here because of the if statement above

kamoltat commented on ceph_devstack/cli.py:
```diff
@@ -44,3 +49,51 @@ async def run():
+    if len(latest_run_subdir) == 0:
+        logger.error("No jobs found!")
+        return 1
+    if len(latest_run_subdir) == 1:
```
can just do job_path = latest_run_subdir[0]. No need to check for 1 run, unless you wanna throw assert

abhi4578 replied:
According the evaluation task, if the number of jobs is greater than 1, then need to display different jobs and user will select one of the job logs to view. 

So, the above condition is put to check if number of jobs equal to one, then no user option needed and if not it goes to below statements which display options -  https://github.com/zmc/ceph-devstack/blob/94f2a8f7b3d87a76e8fac55534b110228eb737df/ceph_devstack/cli.py#L81

abhi4578 replied:
Added comments in the code to make it more clearer

kamoltat commented on ceph_devstack/resources/test/test_logs.py:
```diff
@@ -0,0 +1,101 @@
+import logging
+
+@pytest.mark.parametrize("num_runs,num_jobs,selection",[(0,0,0),(2,0,0),(4,1,0),(4,3,2),(3, 2, 3)])
+@pytest.mark.parametrize("flag_set",["true","false"])
```
Pretty elegant, but please comment on how this i ran 10 times, 5 combinations times 2 combinations

abhi4578 replied:
Sure will add the comments.

kamoltat commented on ceph_devstack/cli.py:
```diff
@@ -2,8 +2,10 @@
 import yaml
-
 from pathlib import Path
+from pydoc import ttypager
```
good choice

kamoltat reviewed (commented):
First round comments for GSOC evaluation, good job! Will not approve the PR since it is just evaluation task.

kamoltat commented:
@abhi4578 also can you please rebase your PR against the main branch thank you

abhi4578 committed:
Update ceph_devstack/resources/test/test_logs.py

Co-authored-by: Kamoltat (Junior) Sirivadhna <ksirivad@gmail.com>
Signed-off-by: Abhi  <29863835+abhi4578@users.noreply.github.com>

abhi4578 committed:
Resolving issues raised in review

- Add appropriate comments for better
  understanding
- Add try, except block while parsing
  latest job directory paths to catch
  ValueError
- Make flag_set to boolean from str
  for simplicity in test_logs.py
- Added chec condition if teuthology.log
  file exists before reading to
  prevent exceptions

abhi4578 committed:
Fix teuthology.log file exists and add test cases

abhi4578 committed:
Replacing test config yaml file with toml file

- Also replace return statements with exit
  statements in test_logs.py file because
  pytest warnings

abhi4578 commented:
@kamoltat I have added commits resolving the above issues and rebased it against main branch. Please review. 

Find the screenshots of 
1. ```ceph-devstack --logs```
![Screenshot from 2025-04-10 21-28-54](https://github.com/user-attachments/assets/8c13d99f-ba23-4ce2-b3e6-3be5e1e06ae2)

2. ```ceph-devstack --logs --log-file```
![Screenshot from 2025-04-10 21-28-10](https://github.com/user-attachments/assets/cc8d5db9-67dc-4a77-9e1c-6a5ba3de9bdc)

3.  Tests ```python3 -m pytest ceph_devstack/resources/test/test_logs.py```
![Screenshot from 2025-04-10 21-26-51](https://github.com/user-attachments/assets/f6271c8c-f09a-4f59-bd3e-411b24db760c)

kamoltat reviewed (approved):
Good work!

abhi4578 commented:
> Good work!

Thank you!
